man/pos2tracery.1
.TH "POS2TRACERY" "1" "November 2020" "v1.0.0" "POS 2 Tracery Help"
.SH "NAME"
\fBpos2tracery\fR
.P
npm version \fIhttps://badge\.fury\.io/js/pos2tracery\.svg\fR \fIhttps://badge\.fury\.io/js/pos2tracery\fR
.QP
.P
pos2tracery is a tool stemming from my
NaNoGenMo 2017 Artbook Project \fIhttp://jkirchartz\.com/NaNoGenMo/2017/\fR and
modified with behaviors from the NaNoGenMo 2018 Wink
Project \fIhttp://jkirchartz\.com/NaNoGenMo/2018/\fR it uses the Wink POS
Tagger \fIhttps://winkjs\.org/wink\-pos\-tagger/\fR which uses the transformation
based learning (TBL) approach to create tracery grammars to take sentence forms
from a corpus, but replace the parts of speech within them with other parts of
speech from throughout the corpus\. It is aware of contractions and punction;
and creates tracery's default english modifiers\.

.
.SH INSTALL
.P
.RS 2
.nf
npm install \-g pos2tracery
.fi
.RE
.SH SYNOPSIS
.QP
.P
pos2tracery currently consists of 3 tools, pos2tracery, merge, and generate
each can be run as a standalone app, or imported into your projects\.

.
.SH CLI USAGE
.SS POS
.P
Generate tracery grammars from POS tags\.
.P
.RS 2
.nf
pos2tracery [pos|p] <input> [output] [options]
example: pos2tracery pos corpus\.txt grammar\.json

Positionals:
  input   input/source file  [string] [required]
  output  optional output/destination file, if not set file prints to stdout  [string]

Options:
  \-\-version        Show version number  [boolean]
  \-\-verbose, \-v    print details while processing  [count]
  \-\-percent, \-p    limit the percentage of words replaced with their POS tags number between 1 and 100  [number] [default: 100]
  \-\-modifiers, \-m  replace english modifiers with their equivalent tracery\.modifier function  [boolean] [default: false]
  \-\-origin, \-o     Include "origin" key in tracery file, specify \-\-no\-origin to not add this key  [boolean] [default: true]
  \-\-ignore, \-i     list of parts of speech to not tagify  [array] [default: []]
  \-\-split, \-s      determine string splitting strategy: line, paragraph, or sentence  [choices: "l", "p", "s"] [default: "s"]
  \-h, \-\-help       Show help  [boolean]
.fi
.RE
.SS Soundex
.P
Generate tracery grammars with Soundex\.
.P
.RS 2
.nf
pos2tracery soundex <input> [output]
example: pos2tracery soundex corpus\.txt grammar\.json

Positionals:
  input   input/source file  [string] [required]
  output  optional output/destination file, if not set file prints to stdout  [string]

Options:
  \-\-version      Show version number  [boolean]
  \-\-verbose, \-v  print details while processing  [count]
  \-\-percent, \-p  limit the percentage of words replaced with their POS tags number between 1 and 100  [number] [default: 100]
  \-\-origin       Include "origin" key in tracery file, specify \-\-no\-origin to not add this key  [boolean] [default: true]
  \-\-split, \-s    determine string splitting strategy: line, paragraph, or sentence  [choices: "l", "p", "s"] [default: "s"]
  \-h, \-\-help     Show help  [boolean]
.fi
.RE
.SS Merge
.P
Merge 2 tracery grammars with
.P
.RS 2
.nf
pos2tracery merge <inputA> <inputB> [output]
example: pos2tracery merge grammar\.json grammar2\.json combined_output\.json

Positionals:
  inputA  input/source file  [string] [required]
  inputB  input/source file  [string] [required]
  output  optional output/destination file, if not set file prints to stdout  [string]

Options:
  \-\-version      Show version number  [boolean]
  \-v, \-\-verbose  print details while processing  [boolean] [default: false]
  \-d, \-\-dupes  [boolean] [default: true]
  \-h, \-\-help     Show help  [boolean]
.fi
.RE
.SS Generate
.P
Generate text from a tracery grammar
.P
.RS 2
.nf
pos2tracery generate <input>
pos2tracery generate grammar\.json

Positionals:
  input  input/source grammar file  [string] [required]

Options:
  \-\-version        Show version number  [boolean]
  \-m, \-\-modifiers  use modifiers  [boolean] [default: true]
  \-o, \-\-origin     use specified origin to create sentences  [string] [default: "origin"]
  \-\-repeat, \-r     define number of sentence to generate  [number] [default: 1]
  \-\-evaluate, \-e   evaluate tracery as javascript template (write javascript inside ${} in tracery)  [boolean] [default: false]
  \-\-verbose, \-v    output information about internal processes  [count]
  \-h, \-\-help       Show help  [boolean]
.fi
.RE
.SS Delete
.P
Delete keys in a tracery grammar through a whitelist and/or a blacklist
.P
.RS 2
.nf
pos2tracery delete <input> [output]
pos2tracery delete grammar\.json grammar\-clean\.json \-t story

Positionals:
  input   input/source file  [string] [required]
  output  optional output/destination file, if not set file prints to stdout  [string]

Options:
  \-\-version      Show version number  [boolean]
  \-\-keep, \-k     a list of keys to keep from the input json file (overrides duplicate values in toss)  [array] [default: []]
  \-\-toss, \-t     a list of keys to delete from the input json file  [array] [default: []]
  \-v, \-\-verbose  print details while processing  [count]
  \-h, \-\-help     Show help  [boolean]
.fi
.RE
.SH MODULE USAGE
.P
pos2tracery can also be used inside of your node projects\. Each option is set by using the long\-form version of the CLI option
The only change is that since \fBdelete\fP is a reserved word in Javascript the function is called \fBdel\fP
.P
.RS 2
.nf
const p2t = require('pos2tracery');

let merged = p2t\.merge({
  inputA: p2t\.del({
            input: "\./grammar_A\.json",
            toss: "story"
          }),
  inputB: p2t\.del({
            input: "\./grammar_B\.json",
            keep: "story"
          })
  });

  p2t\.generate({
    input: merged,
    modifiers: true
  });
.fi
.RE



lib/pos2tracery.js
/*
 * pos2tracery.js
 * Copyleft (ↄ) 2019 jkirchartz <me@jkirchartz.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

const utils = require('./utils');
const util = require('wink-nlp-utils');
const pos = require('wink-pos-tagger');
const contractions = require('expand-contractions');
const tagger = pos();


exports.command = 'pos <input> [output]';
exports.alias = 'p';
exports.describe = 'convert corpus to tracery with POS tags';
exports.builder = function (yargs) {
  yargs.usage('Usage: $0 [pos|p] <input> [output] [options]')
    .positional("input", {
      describe: "input/source file",
      type: "string",
      nargs: 1,
      demand: "input/source file is required"
    }).positional("output", {
      describe: "optional output/destination file, if not set file prints to stdout",
      type: "string",
      nargs: 1,
    }).options({
      "verbose": {
        alias: "v",
        describe: "print details while processing",
        type: "count"
      },
      "percent": {
        alias: "p",
        describe: "limit the percentage of words replaced with their POS tags, as number between 1 and 100",
        default: 100,
        type: "number",
        nargs: 1,
      },
      "modifiers" : {
        alias: "m",
        describe: "replace english modifiers with their equivalent tracery.modifier function",
        default: false,
        type: "boolean"
      },
      "origin": {
        alias: "o",
        describe: "Include \"origin\" key in tracery file, specify --no-origin to not add this key",
        type: "boolean",
        default: true
      },
      "ignore" : {
        "alias": "i",
        describe: "list of parts of speech to not tagify",
        default: [],
        type: "array"
      },
      "split" : {
        alias: "s",
        describe: "determine string splitting strategy: line, paragraph, or sentence",
        choices: ["l", "p", "s"],
        nargs: 1,
        default: "s"
      }
    }).coerce(['input', 'output'], require('path').resolve).help('h').alias('h', 'help');
};

exports.handler = function pos2tracery(args, type) {
  var tracery = { 'sentences' : []};
  if (args.origin) {
    tracery["origin"] = ["#sentences#"];
  }

  const log = utils.logger(args.verbose);

  // generate tracery from POS
  const parseSentence = (str) => {
    // clean up sentence
    str = str.replace(/[\r\n\s]+/g, ' ');
    str = str.replace(/--/g, '-');
    str = contractions.expand(str);
    // tag sentence
    let taggedSentence = tagger.tagSentence(str);
    // fix peculiarity of wink-pos-tagger tagging sentences as NNP
    let scragglers = [];
    taggedSentence = taggedSentence.filter((obj) => {
      if (obj.pos.slice(0,3) === "NPP" &&
        (obj.value.indexOf(" ") > -1 || obj.value.indexOf("\"") > -1)) {
        scragglers.push(tagger.tagSentence(obj.value));
        return false;
      }
      return true;
    });
    taggedSentence = taggedSentence.concat(scragglers);
    // munge sentences to create tracery
    taggedSentence = taggedSentence.filter(tagifySentence);
    log(2)("tagged sentence:", taggedSentence);
    return tidySentences(taggedSentence);
  };

  const tidySentences = (arr) => {
    if (!arr || arr.length == 0) {
      return "";
    }
    // reduce array of objects to one object and return it's value
    return arr.reduce((a, b) => {
      // apply proper spacing, being mindful of punctuation
      if ( b.tag === "punctuation" || b.pos === "POS" ) {
        return { value : a.value + "" + b.value };
      } else {
        return { value : a.value + " " + b.value };
      }
    }).value;
  };

  let was_a_or_an = false;
  const tagifySentence = (obj, i, arr) => {
    let key = obj.pos === "." ? "ending" : obj.pos;
    let word = obj.normal;
    log(3)("found: %s (%s):", obj.value, word);
    if (args.ignore.indexOf(obj.pos) !== -1) {
      return obj; // return ignored values unscathed
    } else if(obj.tag === "punctuation" || obj.pos === "POS") {
      // escape problematic punctuation, just in case
      obj.value = ["'", '"', '[', ']', '#'].indexOf(obj.value) > -1 ? "\\" + obj.value : obj.value;
      return obj;
    }
    if ( ! tracery[key] ) {
      tracery[key] = [];
    }
    if (obj.lemma && args.modifiers) {
      if ((obj.normal.slice(-2) === 'ed' && obj.lemma.slice(-2) !== 'ed')) {
        word = obj.lemma;
      }
    }
    if ( tracery[key].indexOf(word) === -1  && (word !== "a" || word !=="an")) {
      tracery[key].push(word);
    }
    obj.value = utils.chance(args.percent) ? obj.value : '#' + key + '#';
    if (tracery[key].indexOf(word) > -1 ) {
      if(args.modifiers && obj.value.indexOf('#') >= -1) {
        if((i === 0 && obj.pos !== "\"") ||
          (i >= 1 && arr[i-1].pos === "\"")) {
          //capitalize first letter in a sentence
          obj.value = '#' + key + '.capitalize#';
        } else if (obj.pos.slice(0,3) === "NNP") {
          // capitalize all proper nouns
          obj.value = '#' + key + '.capitalizeAll#';
        }
        if (obj.lemma && obj.normal.slice(-2) === 'ed' && obj.lemma.slice(-2) !== 'ed') {
          // preserve "-ed" words
          obj.value = '#' + key + '.ed#';
        }
        if (was_a_or_an) {
          // ensure correct "A" or "An" is used
          obj.value = '#' + key + '.a#';
        }
      }
      log(3)("adding %s to %s (%s)", obj.value, key, obj.pos);
    }
    // don't save one-letter non-words
    if (obj.value.length === 1 &&
      (word !== "a" || word !=="o" || word !== "i")) {
      return false;
    }
    if (word === "a" || word === "an") {
      was_a_or_an = true;
      return false;
    } else {
      was_a_or_an = false;
    }
    return obj;
  };

  const parseCorpus = (file) => {
    let corpus = file.toString();
    // tidy newlines
    corpus = corpus.replace(/\r\n/g, '\n');

    var splitting = "";
    switch (args.split || "s") {
    case "l": // line
      corpus = corpus.split(/\n+/);
      splitting = "line";
      break;
    case "p": // paragraph
      corpus = corpus.split(/\n\n+/);
      splitting = "paragraph";
      break;
    default: // sentence
      corpus = util.string.sentences(corpus.replace(/\n+/, ' ').replace(/\s+/, ' '));
      splitting = "sentence";
    }
    log(1)("splitting on %s", splitting);
    // parse sentences, remove one-character sentences
    corpus = corpus.map(parseSentence).filter((s) => s.length > 1);
    tracery['sentences'] = corpus;

    utils.write(args.output, tracery);
    return tracery;
  };

  if (args.input) {
    var file = utils.read(args.input);
    if (file.length) {
      if(type === "module") {
        return parseCorpus(file);
      } else {
        parseCorpus(file);
      }
    } else {
      log(0)(file.error);
    }
  }
};

